queryable-spec.ts

```diff
@@ -0,0 +1,421 @@
+ *****************************************************************************/
+
+/*
+ * Map-like representation of Bindings as using plain objects could lead
+ * to collisions between variable names and object properties. Support for
+ * immutability is required (but implementations are free to be mutable) which
+ * determines the return value of the set() and delete() methods to be an 
+ * instance of Bindings (potentially a different one).
+ */ 
+interface Bindings {
+  type: 'bindings';
+  has: (key: RDF.Variable) => boolean;
+  get: (key: RDF.Variable) => RDF.Term | undefined;
+  set: (key: RDF.Variable, value: RDF.Term) => Bindings;
+  delete: (key: RDF.Variable) => Bindings;
+  keys: () => Iterator<RDF.Variable>;
+  values: () => Iterator<RDF.Term>;
+  entries: () => Iterator<[RDF.Variable, RDF.Term]>;
+  forEach: (fn: (value: RDF.Term, key: RDF.Variable) => any) => void;
+  size: number;
+}
+
+/*
+ * Bindings objects are created using a dedicated factory, keeping in line
+ * with DataFactory.quad(). This also helps with facilitating support for
+ * immutability. Basic helper methods must also be provided for the most 
+ * common manipulations of bindings objects.
+ */
+interface BindingsFactory {
+  bindings: (entries?: [RDF.Variable, RDF.Term][]) => Bindings;
+  filter: (bindings: Bindings, fn: (value: RDF.Term, key: RDF.Variable) => boolean) => Bindings;
+  map: (bindings: Bindings, fn: (value: RDF.Term, key: RDF.Variable) => RDF.Term) => Bindings;
+  merge: (left: Bindings, right: Bindings) => Bindings;
+  mergeWith: (
+    merger: (left: RDF.Term, right: RDF.Term, key: RDF.Variable) => RDF.Term,
+    left: Bindings,
+    right: Bindings,
+  ) => Bindings;
+}
+
+/*
+ * Base interfaces to represent query results. These interfaces are generic
+ * with respect to the types of query metadata objects. These can be extended
+ * by implementors.
+ */
+
+interface QueryableResultMetadata<OrderItemsType extends TermName | RDF.Variable> {
+  cardinality?: number;
+  availableOrders?: QueryOperationOrder<OrderItemsType>[];
+  [key: string]: any;
+}
+
+interface BaseQueryableResult {
+  type: 'bindings' | 'quads' | 'boolean' | 'void';
+}
+
+interface QueryableResultBindings extends BaseQueryableResult {
+  type: 'bindings';
+  bindings(opts?: { order?: QueryOperationOrder<RDF.Variable> }): Stream<Bindings>;
+  variables: RDF.Variable[];
+  metadata(opts: { [key: string]: any }): Promise<QueryableResultMetadata<RDF.Variable>>;
+}
+    
+interface QueryableResultQuads extends BaseQueryableResult {
+  type: 'quads';
+  quads(opts?: { order?: QueryOperationOrder<TermName> }): Stream<RDF.Quad>;
+  metadata(opts: { [key: string]: any }): Promise<QueryableResultMetadata<TermName>>;
+}
+
+interface QueryableResultBoolean extends BaseQueryableResult {
+  type: 'boolean';
+  value(): Promise<boolean>;
+}
+
+interface QueryableResultVoid extends BaseQueryableResult {
+  type: 'void';
+  execute(): Promise<void>;
```

Comment:
@rubensworks @gsvarovsky I like how both `Filterable` and `Queryable` are converting towards the same general pattern. I think renaming result-specific methods such as `.bindings()`, `.quads()`, .`value()` and `.void()` to `execute()` could add cohesiveness and facilitate understanding of the pattern. What do you think?

Reply:
Hmm, I like that idea. We already have enough information to distinguish the different result types using the `type` field, so I don't see a strong reason for having the differently named `.bindings()`, `.quads()`, `.value()` methods.

We do have to make sure that it's clear that the signature may not be 100% identical, e.g. for quads in the Filterable source: `execute(opts?: { order?: QueryOperationOrder<TermName> })`